samples/nrf9160/serial_lte_modem/src/slm_at_ftp.c

```diff
@@ -0,0 +1,520 @@
+
+	/* Parse AT command */
+	param_count = at_params_valid_count_get(&at_param_list);
+	if (param_count < 3) {
+		return -EINVAL;
+	}
+	ret = at_params_string_get(&at_param_list, 2, folder, &sz_folder);
+	if (ret) {
+		return ret;
+	}
+	folder[sz_folder] = '\0';
+
+	ret = ftp_rmd(folder);
+	return (ret == FTP_CODE_250)?0 : -1;
+}
+
+/* AT#XFTP="rename",<file_old>,<file_new> */
+static int do_ftp_rename(void)
+{
+	int ret;
+	char file_old[128];
+	int sz_file_old = 128;
+	char file_new[128];
+	int sz_file_new = 128;
+	int param_count;
+
+	/* Parse AT command */
+	param_count = at_params_valid_count_get(&at_param_list);
+	if (param_count < 4) {
+		return -EINVAL;
+	}
+	ret = at_params_string_get(&at_param_list, 2, file_old, &sz_file_old);
+	if (ret) {
+		return ret;
+	}
+	file_old[sz_file_old] = '\0';
+	ret = at_params_string_get(&at_param_list, 3, file_new, &sz_file_new);
+	if (ret) {
+		return ret;
+	}
+	file_new[sz_file_new] = '\0';
+
+	ret = ftp_rename(file_old, file_new);
+	return (ret == FTP_CODE_250)?0 : -1;
+}
+
+/* AT#XFTP="delete",<file> */
+static int do_ftp_delete(void)
+{
+	int ret;
+	char file[128];
+	int sz_file = 128;
+	int param_count;
+
+	/* Parse AT command */
+	param_count = at_params_valid_count_get(&at_param_list);
+	if (param_count < 3) {
+		return -EINVAL;
+	}
+	ret = at_params_string_get(&at_param_list, 2, file, &sz_file);
+	if (ret) {
+		return ret;
+	}
+	file[sz_file] = '\0';
+
+	ret = ftp_delete(file);
+	return (ret == FTP_CODE_250)?0 : -1;
+}
+
+/* AT#XFTP="get",<file> */
+static int do_ftp_get(void)
+{
+	int ret;
+	char file[128];
+	int sz_file = 128;
+	int param_count;
+
+	/* Parse AT command */
+	param_count = at_params_valid_count_get(&at_param_list);
+	if (param_count < 3) {
+		return -EINVAL;
+	}
+	ret = at_params_string_get(&at_param_list, 2, file, &sz_file);
+	if (ret) {
+		return ret;
+	}
+	file[sz_file] = '\0';
+
+	ret = ftp_get(file);
+	return (ret == FTP_CODE_226)?0 : -1;
+}
+
+/* AT#XFTP="put",<file>[<datatype>,<data>] */
+static int do_ftp_put(void)
+{
+	int ret;
+	char file[128];
+	int sz_file = 128;
+	int param_count;
+
+	/* Parse AT command */
+	param_count = at_params_valid_count_get(&at_param_list);
+	if (param_count < 3) {
+		return -EINVAL;
+	}
+	ret = at_params_string_get(&at_param_list, 2, file, &sz_file);
+	if (ret) {
+		return ret;
+	}
+	file[sz_file] = '\0';
+
+	if (param_count > 4) {
+		u16_t type;
+		char data[NET_IPV4_MTU];
+		int size;
+
+		ret = at_params_short_get(&at_param_list, 3, &type);
+		if (ret) {
+			return ret;
+		}
+		size = NET_IPV4_MTU;
+		ret = at_params_string_get(&at_param_list, 4, data, &size);
+		if (ret) {
+			return ret;
+		}
+		if (type == DATATYPE_HEXADECIMAL) {
+			u8_t data_hex[size / 2];
+
+			ret = slm_util_atoh(data, size, data_hex, size / 2);
+			if (ret > 0) {
+				ret = ftp_put(file, data_hex, ret);
+			}
+		} else {
+			ret = ftp_put(file, data, size);
+		}
+	} else {
+		ret = ftp_put(file, NULL, 0);
+	}
+
+	return (ret == FTP_CODE_226)?0 : -1;
```

Comment:
I wonder if returning an FTP result code is actually a good idea, the user of this FTP client library needs to know which result code to expect for each operation. Wouldn't it be enough to return 0 on success, and check the result code internally in the API functions?

Reply:
Yes we can add this logic to FTP client in future commit.